test/e2e/gameserver_test.go

```diff
@@ -213,6 +213,96 @@ func TestGameServerUnhealthyAfterDeletingPod(t *testing.T) {
 	assert.NoError(t, err)
 }
 
+func TestGameServerRestartBeforeReadyCrash(t *testing.T) {
+	t.Parallel()
+	logger := logrus.WithField("test", t.Name())
+
+	gs := defaultGameServer(defaultNs)
+	// give some buffer with gameservers crashing and coming back
+	gs.Spec.Health.PeriodSeconds = 30
+	gs.Spec.Template.Spec.Containers[0].Env = append(gs.Spec.Template.Spec.Containers[0].Env, corev1.EnvVar{Name: "READY", Value: "FALSE"})
+	gsClient := framework.AgonesClient.AgonesV1().GameServers(defaultNs)
+	newGs, err := gsClient.Create(gs)
+	if !assert.NoError(t, err) {
+		assert.Fail(t, "could not create the gameserver")
+	}
+
+	defer gsClient.Delete(newGs.ObjectMeta.Name, nil) // nolint: errcheck
+
+	logger.Info("Waiting for us to have an address to send things to")
+	newGs, err = framework.WaitForGameServerState(newGs, agonesv1.GameServerStateScheduled, time.Minute)
+	assert.NoError(t, err)
+
+	// crash the pod
+	conn, err := net.Dial("udp", fmt.Sprintf("%s:%d", newGs.Status.Address, newGs.Status.Ports[0].Port))
+	assert.NoError(t, err)
+	defer conn.Close() // nolint: errcheck
+
+	logger.Info("crashing, and waiting to see restart")
+	err = wait.Poll(time.Second, 3*time.Minute, func() (bool, error) {
+		// not sure when the pod process will actually start, so lets send a crash every time, just in case
+		// process may not be there, so we don't care about the error
+		conn.Write([]byte("CRASH")) // nolint: errcheck
+		pod, err := framework.KubeClient.CoreV1().Pods(defaultNs).Get(newGs.ObjectMeta.Name, metav1.GetOptions{})
+		if err != nil {
+			return true, err
```

Comment:
should this return false to retry the loop?

Reply:
So we created the GameServer previously, and waited for it to be `Scheduled`. If the Pod isn't there, then something has gone horribly wrong, since `Scheduled` state indicates that a Pod exists in some form or another.

Reply:
Got it. So the idea is to break the loop and return an error to let the error bubble up.